classes/webservice/ApiNode.php

```diff
@@ -0,0 +1,378 @@
+     */
+    public function addLanguageNode(string $name): self
+    {
+        $newNode = self::lang($name);
+        $this->nodes[] = $newNode;
+
+        return $newNode;
+    }
+
+    /**
+     * Create new ApiNode of type "parent" and appends it as child to current ApiNode
+     *
+     * @param string $name
+     * @param array $attributes
+     *
+     * @return ApiNode Created child node
+     */
+    public function addParentNode(string $name = null, array $attributes = []): self
+    {
+        $newNode = self::parent($name, $attributes);
+        $this->nodes[] = $newNode;
+
+        return $newNode;
+    }
+
+    /**
+     * Create new ApiNode of type "list" and appends it as child to current ApiNode
+     *
+     * @param string $name
+     * @param array $attributes
+     *
+     * @return ApiNode Created child node
+     */
+    public function addListNode(string $name = null, array $attributes = [])
+    {
+        $newNode = self::list($name, $attributes);
+        $this->nodes[] = $newNode;
+
+        return $newNode;
+    }
+
+    /**
+     * Transform $field array into ApiNode and appends it as child to current node
+     *
+     * @param array $field
+     * @param string|null $schemaToDisplay
+     *
+     * @return self
+     */
+    public function addField(array $field, ?string $schemaToDisplay = null): self
+    {
+        $newNode = self::value($field['sqlId']);
+
+        if (isset($field['encode'])) {
+            $newNode->addAttribute('encode', $field['encode']);
+        }
+
+        if (!empty($field['synopsis_details']) && $schemaToDisplay !== 'blank') {
+            foreach ($field['synopsis_details'] as $name => $detail) {
+                $newNode->addAttribute($name, is_array($detail) ? implode(' ', $detail) : $detail);
+            }
+        }
+
+        // display i18n fields
+        if (!empty($field['i18n'])) {
+            foreach (self::$languages as $language) {
+                $langAttributes = ['id' => $language];
+
+                if (isset($field['synopsis_details']) || (isset($field['value']) && is_array($field['value']))) {
+                    $langAttributes['xlink:href'] = WebserviceOutputBuilderCore::$wsUrl . 'languages/' . $language;
+                    if (isset($field['synopsis_details']) && $schemaToDisplay != 'blank') {
+                        $langAttributes['format'] = 'isUnsignedId';
+                    }
+                }
+
+                $newNode->setType(self::TYPE_LANGUAGE);
+                $newNode->addNode('language', $field['value'][$language] ?? '')
+                    ->setAttributes($langAttributes);
+            }
+        } else {
+            // display not i18n fields value
+            if (array_key_exists('xlink_resource', $field) && $schemaToDisplay != 'blank') {
+                if (!is_array($field['xlink_resource'])) {
+                    $xlink = WebserviceOutputBuilderCore::$wsUrl . $field['xlink_resource'] . '/' . $field['value'];
```

Comment:
In this block, you don't test if isset($field['value'])

Reply:
Added in accca16